### ISSUE TITLE
Apicurio client api not backwards compatible to older releases (Timzeone parsing error)

### ISSUE THREAD

**Issue by davidvoit** [Apicurio/apicurio-registry#3024]
Hi Apicurio-Team,

## Description

**Registry
Version**:  Server: 2.3.1 Client: 2.4.0
**Persistence type**: sql

A kafka connect client can't access the schema-registry with the 2.4.0 package

Raw json output:
```
{"name":"XXX","createdBy":"","createdOn":"2021-07-12T07:30:06+0000","modifiedBy":"","modifiedOn":"2021-07-12T07:30:06+0000","id":"XXX-value","version":"1","type":"AVRO","globalId":26,"state":"ENABLED","contentId":25,"references":[]}
```

```
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.util.Date` from String "2021-07-12T07:30:06+0000": expected format "yyyy-MM-dd'T'HH:mm:ss'Z'"
at [Source: (jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream); line: 1, column: 44] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["createdOn"])
    at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67)
    at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:1991)
    at com.fasterxml.jackson.databind.DeserializationContext.handleWeirdStringValue(DeserializationContext.java:1219)
    at com.fasterxml.jackson.databind.deser.std.DateDeserializers$DateBasedDeserializer._parseDate(DateDeserializers.java:195)
    at com.fasterxml.jackson.databind.deser.std.DateDeserializers$DateDeserializer.deserialize(DateDeserializers.java:303)
    at com.fasterxml.jackson.databind.deser.std.DateDeserializers$DateDeserializer.deserialize(DateDeserializers.java:281)
    at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:313)
    at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:176)
    at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
    at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4674)
    at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3674)
    at io.apicurio.rest.client.handler.BodyHandler.lambda$toSupplierOfType$1(BodyHandler.java:56)
    ... 30 more
```

A 2.3.1 client works fine.

Reason: 

Fields in (generated) ArtificatMetadata have now the following pattern:
```
@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss'Z'", timezone = "UTC")
@JsonProperty("createdOn")
private Date createdOn;
```

in version 2.3.1 it is:
```
@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssZ", timezone = "UTC")
@JsonProperty("createdOn")
private Date createdOn;
```

Possible fix(?):
 apicurio-codegen

```
diff --git a/core/src/main/java/io/apicurio/hub/api/codegen/pre/OpenApiDateTimeSimpleTypeProcessor.java b/core/src/main/java/io/apicurio/hub/api/codegen/pre/OpenApiDateTimeSimpleTypeProcessor.java
index 7167f72..65d1312 100755
--- a/core/src/main/java/io/apicurio/hub/api/codegen/pre/OpenApiDateTimeSimpleTypeProcessor.java
+++ b/core/src/main/java/io/apicurio/hub/api/codegen/pre/OpenApiDateTimeSimpleTypeProcessor.java
@@ -40,7 +40,7 @@ public class OpenApiDateTimeSimpleTypeProcessor extends CombinedVisitorAdapter {
         OasSchema schema = (OasSchema) node;
         // Switch from int64 format to utc-millisec so that jsonschema2pojo will generate a Long instead of an Integer
         if ("string".equals(schema.type) && "date-time".equals(schema.format)) {
-            String formatPattern = "yyyy-MM-dd'T'HH:mm:ss'Z'";
+            String formatPattern = "yyyy-MM-dd'T'HH:mm:ssX";
             Extension ext = schema.getExtension(CodegenExtensions.FORMAT_PATTERN);
             if (ext != null && ext.value != null) {
                 formatPattern = ext.value.toString();
```

This would produce the requested Z timezone letter in the output but will parse also other timezone fields

Kind Regards,
David

**Comment by carlesarnal** [comment]
cc @andreaTP

**Comment by EricWittmann** [comment]
You are correct.  Sorry about this inconvenience.  I have confirmed locally that the 2.4.0.Final client will fail against older versions of the server.  We'll address this in 2.4.1.Final (so that the 2.4.1.Final client will work with previous versions of the server).

**Comment by ajborley** [comment]
I see what looks like the exact same failure with older clients (2.3.1.Final) talking to the latest server (2.4.0.Final) - this failure is seen at the client side:

```
[kafka-producer-network-thread | kafka-demo-producer--90] INFO org.apache.kafka.clients.Metadata - [Producer clientId=kafka-demo-producer--90] Cluster ID: C1R7qy2NQbuJAEDknxMW1A
java.io.UncheckedIOException: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.util.Date` from String "2022-12-13T20:10:29Z": expected format "yyyy-MM-dd'T'HH:mm:ssZ"
 at [Source: (jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream); line: 1, column: 54] (through reference chain: io.apicurio.registry.rest.v2.beans.ArtifactMetaData["createdOn"])
        at io.apicurio.rest.client.handler.BodyHandler.lambda$toSupplierOfType$1(BodyHandler.java:60)
        at io.apicurio.rest.client.JdkHttpClient.sendRequest(JdkHttpClient.java:204)
        at io.apicurio.registry.rest.client.impl.RegistryClientImpl.createArtifact(RegistryClientImpl.java:263)
        at io.apicurio.registry.rest.client.RegistryClient.createArtifact(RegistryClient.java:134)
        at io.apicurio.registry.resolver.DefaultSchemaResolver.lambda$handleAutoCreateArtifact$2(DefaultSchemaResolver.java:236)
        at io.apicurio.registry.resolver.ERCache.lambda$getValue$0(ERCache.java:142)
        at io.apicurio.registry.resolver.ERCache.retry(ERCache.java:181)
        at io.apicurio.registry.resolver.ERCache.getValue(ERCache.java:141)
        at io.apicurio.registry.resolver.ERCache.getByContent(ERCache.java:121)
        at io.apicurio.registry.resolver.DefaultSchemaResolver.handleAutoCreateArtifact(DefaultSchemaResolver.java:234)
        at io.apicurio.registry.resolver.DefaultSchemaResolver.getSchemaFromRegistry(DefaultSchemaResolver.java:115)
        at io.apicurio.registry.resolver.DefaultSchemaResolver.resolveSchema(DefaultSchemaResolver.java:88)
        at io.apicurio.registry.serde.AbstractKafkaSerializer.serialize(AbstractKafkaSerializer.java:83)
        at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:929)
        at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:889)
        at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:775)

```

Will the fix at https://github.com/Apicurio/apicurio-common-rest-client/pull/116 also work in this case? Or must we update clients to 2.4.0.Final?

**Comment by EricWittmann** [comment]
You are right, @ajborley - old clients will fail against the new server.  You'll need to upgrade clients first, then server.  We could add a feature to 2.4.1.Final (server) to allow it to be configured to use the old date format if that would help with upgrades.

@andreaTP WDYT?

**Comment by andreaTP** [comment]
@EricWittmann with your latest fix the clients are going to be backward compatible, so, no issue on that side I believe.

Regarding the server I think that we should not make it configurable for a few reasons:
- it's breaking the OpenAPI specification
- the date format was not consistent across the codebase and we would need to punctually apply or not the change

**Comment by EricWittmann** [comment]
OK the client should be fixed in 2.4.1.Final (releasing right now).  I'm closing this as a result, but please re-open if it doesn't work as expected.

PR that fixes this issue is here:  https://github.com/Apicurio/apicurio-common-rest-client/pull/116
